gix-discover/src/repository.rs

```diff
@@ -112,8 +114,11 @@ pub enum Kind {
     /// Note that this is merely a guess at this point as we didn't read the configuration yet.
     ///
     /// Also note that due to optimizing for performance and *just* making an educated *guess in some situations*,
-    /// we may consider a non-bare repository bare if it it doesn't have an index yet due to be freshly initialized.
-    /// The caller is has to handle this, typically by reading the configuration.
+    /// we may consider a non-bare repository bare if it doesn't have an index yet due to be freshly initialized.
+    /// The caller has to handle this, typically by reading the configuration.
+    ///
+    /// It could also be a directory which is non-bare by configuration, but is *not* named `.git`.
+    /// Unusual, but it's possible that a worktree is configured via `core.worktree`.
```

Comment:
Is there also the possibility that the worktree path is established by the presence of a [`GIT_WORK_TREE`](https://git-scm.com/docs/git#Documentation/git.txt-codeGITWORKTREEcode) environment variable? (Or by the presence of a [`GIT_DIR`](https://git-scm.com/docs/git#Documentation/git.txt-codeGITDIRcode) environment variable together with no `GIT_WORK_TREE` variable? In this case, I think the worktree is implicitly taken to be the process CWD, at least in Git.)

Reply:
In `gitoxide`, plumbing crates aren't allowed to read environment variables directly, so this is not handled here.
There is [a function in `gix`](https://github.com/Gitoxide-Labs/gitoxide/blob/f8952e4cbfaf9ab7ddc12a028a1cdb821ac9a3b1/gix/src/open/repository.rs#L114) though that reads these variables and uses them detect repositories from there. 
It's most definitely not fully conforming with what Git does.

Reply:
Thanks--I've commented https://github.com/GitoxideLabs/gitoxide/pull/1585#discussion_r1797734926 with a note about including or otherwise revisiting this when I open the other environment-related compatibility issues per the plan in https://github.com/GitoxideLabs/gitoxide/pull/1585#discussion_r1764341739.

Reply:
Maybe as an additional note, despite claiming it's 'most definitely' not conforming, what I tried to say is that I didn't really try to make it conforming in the first place. To me it felt like environment variables are only needed in certain contexts, for instance, when creating a Git hook with `gitoxide`, and the implementation should rather be so that `gitoxide` can be used in certain situations (of which I probably only know one), and otherwise never be able to lead to something surprising. This is a general problem with environment variables, their implicitness, and `gitoxide` already makes them explicit by [mapping every one of them to a git configuration variable](https://github.com/GitoxideLabs/gitoxide/blob/d2ae9d5f11be9f2561f6799d88804d0d8eae33ef/gix/src/config/cache/init.rs#L311), which is used in its place.